New action to update a subscription's next payment date

kuldeep3 opened the pull request:
**Changes proposed in this Pull Request**

1. Action to change subscription next payment date.
2. Create an abstract subscription action that can be extend for any action on subscription.

**How to test the changes in this Pull Request:**

1. Add a manual workflow for the subscription.
1. Choose a rule from the Rule Option.
1. Select `Update Next Payment` option in the `Action` tab.
1. Choose a `date` and `time` for the next payment date.
1. Save the workflow and hit `run` to start the workflow.

**Changelog entry**
 Action to update a subscription's next payment date

kuldeep3 committed:
update-next-payment method

pokhiii commented on includes/abstracts/actions/subscription.php:
```diff
@@ -0,0 +1,38 @@
+ * Abstract class to define action on a subscription object.
+ *
+ * @class Abstract_Action_Subscription
+ * @since 1.0
```
```suggestion
 * @since 1.2.2
```

pokhiii commented on includes/actions/update-next-payment.php:
```diff
@@ -0,0 +1,66 @@
+ *
+ *
+ * @class Action_Subscription_Update_Next_Payment
+ * @since 1.0.0
```
```suggestion
 * @since 1.2.2
```

pokhiii commented on includes/abstracts/actions/subscription.php:
```diff
@@ -0,0 +1,38 @@
+		return $this->workflow->data_layer()->get_subscription();
+	}
+
+}
```
Leave a blank line after this.

pokhiii commented on includes/actions/update-next-payment.php:
```diff
@@ -0,0 +1,66 @@
+		$time->set_title( __( 'New Payment Time', 'automatewoo-subscriptions' ) );
+		$this->add_field( $time );
+	}
+}
```
Leave a blank line after this.

pokhiii commented on includes/actions/update-next-payment.php:
```diff
@@ -0,0 +1,66 @@
+		$subscription     = $this->get_subscription_to_edit();
+		$old_payment_date = $subscription->get_date( 'next_payment' );
+		$subscription->update_meta_data( '_old_schedule_next_payment', $old_payment_date );
+		$subscription->save();
```
Let's not save the original schedule next payment in the meta. Create a private class variable `$original_payment_date` which can be used when adding the subscription note.
```suggestion
		$subscription                            = $this->get_subscription_to_edit();
		$this->original_payment_date = $subscription->get_date( 'next_payment' );
```

kuldeep3 committed:
added note for subscription renewal date modification

pokhiii commented on includes/actions/update-next-payment.php:
```diff
@@ -0,0 +1,81 @@
+	 * @throws \Exception When there is an error.
+	 */
+	public function edit_subscription( $subscription ) {
+		$old_payment_date = $subscription->get_date( 'next_payment' );
+		$subscription->update_meta_data( '_old_schedule_next_payment', $old_payment_date );
+		$subscription->save();
```
```suggestion
		$this->original_payment_date = $subscription->get_date( 'next_payment' );
```

pokhiii commented on includes/actions/update-next-payment.php:
```diff
@@ -0,0 +1,81 @@
+	 * @return string
+	 */
+	protected function get_note( $subscription ) {
+		return sprintf( __( '%1$s workflow run: updated next payment date on subscription from %2$s to %3$s', 'automatewoo-subscriptions' ), $this->workflow->get_title(), $subscription->get_meta( '_old_schedule_next_payment' ), $subscription->get_date( 'next_payment' ) );
```
```suggestion
		return sprintf(
			__( '%1$s workflow run: updated next payment date on subscription from %2$s to %3$s', 'automatewoo-subscriptions' ),
			$this->workflow->get_title(),
			$this->original_payment_date,
			$subscription->get_date( 'next_payment' )
		);
```

kuldeep3 committed:
added variable to store the value

kuldeep3 committed:
modified function to show notes

pokhiii committed:
Update update-next-payment.php

pokhiii committed:
Variable names improved

- Variable `new_payment_date_string` renamed to `new_payment_date`
- Method `get_new_payment_date` now returns a DateTime object instead of a date string

abhishek-pokhriyal commented:
Hey @danielbitzer, can you please take a look at this PR and see if this makes sense to be included in the upcoming release?

danielbitzer commented:
Thanks for the contribution @abhishek-pokhriyal and @kuldeep3! I'll review it this week.

abhishek-pokhriyal commented:
Hey @danielbitzer, just checking if you got a chance to review the PR.

danielbitzer commented:
Apologies! A few things came up. It's still on my list, aiming for early next week.

danielbitzer commented:
Apologies for the delay, I've been unexpectedly off work for a few days. It's still on my list!

danbitzer commented on includes/actions/update-next-payment.php:
```diff
@@ -0,0 +1,92 @@
+	 */
+	public function load_admin_details() {
+		parent::load_admin_details();
+		$this->title       = __( 'Update Next Payment', 'automatewoo-subscriptions' );
```
```suggestion
		$this->title       = __( 'Update Next Payment Date', 'automatewoo-subscriptions' );
```
I think 'Date' need to be included in the title

pokhiii replied:
Fixed in [`b1a0cd1`](https://github.com/woocommerce/automatewoo-subscriptions/pull/59/commits/b1a0cd1fdd453712bbbab10895e009e6096d05f1).

danbitzer commented on includes/actions/update-next-payment.php:
```diff
@@ -0,0 +1,92 @@
+ * @class Action_Subscription_Update_Next_Payment
+ * @since 1.2.2
+ */
+class Action_Subscription_Update_Next_Payment extends Abstract_Action_Subscription {
```
```suggestion
class Action_Subscription_Update_Next_Payment_Date extends Abstract_Action_Subscription {
```

pokhiii replied:
Fixed in [`b1a0cd1`](https://github.com/woocommerce/automatewoo-subscriptions/pull/59/commits/b1a0cd1fdd453712bbbab10895e009e6096d05f1).

danbitzer commented on includes/actions/update-next-payment.php:
```diff
@@ -0,0 +1,92 @@
+		$subscription->update_dates( array( 'next_payment' => wcs_get_datetime_utc_string( $new_payment_date ) ) );
+	}
+
+	function load_fields() {
+		$this->load_subscription_renewal_fields();
+	}
```
Is there any reason to have `load_subscription_renewal_fields()` has a separate method here? Seems like we could just add the fields directly in `load_fields()`

pokhiii replied:
Makes sense. Fixed in [`b1a0cd1`](https://github.com/woocommerce/automatewoo-subscriptions/pull/59/commits/b1a0cd1fdd453712bbbab10895e009e6096d05f1).

danbitzer commented on includes/actions/update-next-payment.php:
```diff
@@ -0,0 +1,92 @@
+	 * @throws \Exception When there is an error.
+	 */
+	public function edit_subscription( $subscription ) {
+		$this->original_payment_date = $subscription->get_date( 'next_payment' );
```
I'd prefer if we didn't need to store old payment date as a class property. It seems like the structure in `Abstract_Action_Subscription` class could be simplified to avoid having to store as a prop. I can see that `Abstract_Action_Subscription` is based on the pre-existing `\AutomateWoo\Action_Subscription_Edit_Item_Abstract` class but I'm happy to use a different structure here. I think moving the `run()` method from the parent into the child class could help simplify things.

pokhiii replied:
> I'd prefer if we didn't need to store old payment date as a class property.

I've just removed the class property –– we were using it to specify the old payment date on the subscription notes (not doing that anymore).


>  I can see that `Abstract_Action_Subscription` is based on the pre-existing `\AutomateWoo\Action_Subscription_Edit_Item_Abstract` class

Ahh, I got trapped by the name of the class (`Action_Subscription_Edit_Item_Abstract`) –– I thought it's used to update the subscription items (line/coupon/shipping).


> I think moving the run() method from the parent into the child class could help simplify things.

Not going that route for now.

<hr>

Fixed in [`b1a0cd1`](https://github.com/woocommerce/automatewoo-subscriptions/pull/59/commits/b1a0cd1fdd453712bbbab10895e009e6096d05f1).

<hr>

On a side note, the only thing that I'm not liking about this is the timezone and the format of the new payment date shown on the subscription note. The timezone is UTC and the format is `mysql`. Honestly, it confused me when I was looking at the subscription notes on a real WooCommerce store on which we ran this. Let me know what you think about this.

![image](https://user-images.githubusercontent.com/11808845/102893773-20b66f00-4488-11eb-80f9-0e6b2eb39143.png)

danbitzer replied:
> On a side note, the only thing that I'm not liking about this is the timezone and the format of the new payment date shown on the subscription note

I agree, it would be better displayed in the site's timezone and formatted with `wc_date_format()` and `wc_time_format()`.

danielbitzer reviewed (changes requested):
This is working well in my tests! I've left some inline comments.

pokhiii committed:
Changed code structure and other fixes

1. Renamed action `Update_Next_Payment` to `Update_Next_Payment_Date`
2. Changed class `Update_Next_Payment_Date` to extend `Action_Subscription_Edit_Item_Abstract` instead of `Abstract_Action_Subscription`
3. Get rid of unnecessary class `Abstract_Action_Subscription`
4. Get rid of unnecessary method `load_subscription_renewal_fields`
5. Fix doc comments for methods

abhishek-pokhriyal commented:
@danielbitzer We have implemented the requested changes in the commit [`b1a0cd1`](https://github.com/woocommerce/automatewoo-subscriptions/pull/59/commits/b1a0cd1fdd453712bbbab10895e009e6096d05f1). Please re-review when you get a chance.

Thanks

kuldeep3 commented:
Hey @danielbitzer, just checking if you got a chance to review the PR.

danielbitzer commented:
Sorry for the delay, I've just looked over this again. It's looking good! The recent changes have simplified things a lot so thanks for that!

I've just left one response comment about the date format.

danielbitzer commented:
Hey @abhishek-pokhriyal and @kuldeep3, I'm going to merge this and I'll open an issue to tweak that date formatting.
